File: clients/Ruby/model/server_message.rb
require_relative 'player_view'
require_relative 'player_view'
class ServerMessage
    def self.read_from(stream)
        tag = stream.read_int()
        if tag == ServerMessage::GetAction::TAG
            return ServerMessage::GetAction.read_from(stream)
        end
        if tag == ServerMessage::Finish::TAG
            return ServerMessage::Finish.read_from(stream)
        end
        if tag == ServerMessage::DebugUpdate::TAG
            return ServerMessage::DebugUpdate.read_from(stream)
        end
        raise "Unexpected tag value"
    end

    class GetAction
        TAG = 0
        attr_accessor :player_view
        attr_accessor :debug_available
        def initialize(player_view, debug_available)
            @player_view = player_view
            @debug_available = debug_available
        end
        def self.read_from(stream)
            player_view = PlayerView.read_from(stream)
            debug_available = stream.read_bool()
            GetAction.new(player_view, debug_available)
        end
        def write_to(stream)
            stream.write_int(TAG)
            @player_view.write_to(stream)
            stream.write_bool(@debug_available)
        end
    end
    class Finish
        TAG = 1
        def initialize()
        end
        def self.read_from(stream)
            Finish.new()
        end
        def write_to(stream)
            stream.write_int(TAG)
        end
    end
    class DebugUpdate
        TAG = 2
        attr_accessor :player_view
        def initialize(player_view)
            @player_view = player_view
        end
        def self.read_from(stream)
            player_view = PlayerView.read_from(stream)
            DebugUpdate.new(player_view)
        end
        def write_to(stream)
            stream.write_int(TAG)
            @player_view.write_to(stream)
        end
    end
end

File: clients/JavaScript/model/server-message.js
const PlayerView = require('./player-view');
class ServerMessage {
    static async readFrom(stream) {
        let tag = await stream.readInt();
        if (tag == GetAction.TAG) {
            return await GetAction.readFrom(stream);
        }
        if (tag == Finish.TAG) {
            return await Finish.readFrom(stream);
        }
        if (tag == DebugUpdate.TAG) {
            return await DebugUpdate.readFrom(stream);
        }
        throw new Error("Unexpected tag value");
    }
}

class GetAction extends ServerMessage {
    constructor(playerView, debugAvailable) {
        super();
        this.playerView = playerView;
        this.debugAvailable = debugAvailable;
    }
    static async readFrom(stream) {
        let playerView;
        playerView = await PlayerView.readFrom(stream);
        let debugAvailable;
        debugAvailable = await stream.readBool();
        return new GetAction(playerView, debugAvailable);
    }
    async writeTo(stream) {
        await stream.writeInt(GetAction.TAG);
        let playerView = this.playerView;
        await playerView.writeTo(stream);
        let debugAvailable = this.debugAvailable;
        await stream.writeBool(debugAvailable);
    }
}
ServerMessage.GetAction = GetAction;
GetAction.TAG = 0;
class Finish extends ServerMessage {
    constructor() {
        super();
    }
    static async readFrom(stream) {
        return new Finish();
    }
    async writeTo(stream) {
        await stream.writeInt(Finish.TAG);
    }
}
ServerMessage.Finish = Finish;
Finish.TAG = 1;
class DebugUpdate extends ServerMessage {
    constructor(playerView) {
        super();
        this.playerView = playerView;
    }
    static async readFrom(stream) {
        let playerView;
        playerView = await PlayerView.readFrom(stream);
        return new DebugUpdate(playerView);
    }
    async writeTo(stream) {
        await stream.writeInt(DebugUpdate.TAG);
        let playerView = this.playerView;
        await playerView.writeTo(stream);
    }
}
ServerMessage.DebugUpdate = DebugUpdate;
DebugUpdate.TAG = 2;
module.exports = ServerMessage;
